### PR TITLE
[dotnet] future proof the test

### PR DESCRIPTION
**Pull request by stephen-hawley**
Tests fail, especially reflection based tests.
Wouldn't you like it if a test came with a blueprint on how to fix the issue you created? Yes, of course you would. So here it is.

### CODE REVIEW

File: tests/cecil-tests/BlittablePInvokes.cs [hunk 1]
```diff
@@ -10,6 +10,126 @@
 
 namespace Cecil.Tests {
 
+	// Guide to when this test fails.
+	//
+	// What does this test do?
+	// It ensures that every type in a p/invoke is blittalble. This means that
```
Comment (rolfbjarne) [inline tests/cecil-tests/BlittablePInvokes.cs:16]:
```suggestion
	// It ensures that every type in a p/invoke is blittable. This means that
```

File: tests/cecil-tests/BlittablePInvokes.cs [hunk 2]
```diff
@@ -10,6 +10,126 @@
...
+	//    Former typical case:
+	//    extern static void SomePinvoke (string value);
+	//    SomePinvoke (valueStr);
+	//    Fix:
+	//    extern void SomePinvoke (IntPtr value);
+	//    using var valueTransient = new TransientString (valueStr);
+	//    SomePinvoke (valueTransient);
+	//    NOTE: You need to understand the lifespace of the string. If the call
+	//    holds onto the pointer forever instead of making a copy, you can't use
+	//    TransientString. Instead use Marhsal to allocate the string once.
```
Comment (rolfbjarne) [inline tests/cecil-tests/BlittablePInvokes.cs:37]:
```suggestion
	//    TransientString. Instead use Marshal to allocate the string once.
```

**Review by rolfbjarne — APPROVED** [review]

**Review by dalexsoto — APPROVED** [review]

File: tests/cecil-tests/BlittablePInvokes.cs [hunk 3]
```diff
@@ -10,6 +10,126 @@
...
+	// What does this test do?
+	// It ensures that every type in a p/invoke is blittable. This means that
+	//    the number of allowed types gets cut down to a narrow set: pointers and
+	//    scalars.
+	// What are common failing types?
+	//    1. string
+	//    3. delegate
+	//    4. ref/out types
+	//    5. structs with non-blittable types
+	//    6. arrays
```
Comment (haritha-mohan) [inline tests/cecil-tests/BlittablePInvokes.cs:24]: nit
```suggestion
	//    1. string
	//    2. delegate
	//    3. ref/out types
	//    4. structs with non-blittable types
	//    5. arrays
```

File: tests/cecil-tests/BlittablePInvokes.cs [hunk 4]
```diff
@@ -10,6 +10,126 @@
...
+	//    #if NET
+	//        // calling code is now also unsafe because of the & operator
+	//        unsafe {
+	//            AssignSomeCallback (&SomeCallImpl, someHandle);
+	//        }
+	//    #else
+	//        AssignSomeCallback (SomeCallImpl, someHandle);
+	//    #endif
+	//
+	//    4. ref/out types - these are disallowed. Replace them with pointers and
```
Comment (haritha-mohan) [inline tests/cecil-tests/BlittablePInvokes.cs:82]:
```suggestion
	//    3. ref/out types - these are disallowed. Replace them with pointers and
```

File: tests/cecil-tests/BlittablePInvokes.cs [hunk 5]
```diff
@@ -10,6 +10,126 @@
...
+	//        unsafe {
+	//            TPGetCoordinates (this.GetHandle (), &x, &y);
+	//        }
+	//    #else
+	//        TPGetCoordinates (this.GetHandle (), out x, out y);
+	//    #endif
+	//        return new Point (x, y);
+	//    }
+	//
+	//    5. structs with non-blittable types. I don't think I ran into any of these
```
Comment (haritha-mohan) [inline tests/cecil-tests/BlittablePInvokes.cs:110]:
```suggestion
	//    4. structs with non-blittable types. I don't think I ran into any of these
```

File: tests/cecil-tests/BlittablePInvokes.cs [hunk 6]
```diff
@@ -10,6 +10,126 @@
...
+	//       *except* for structs with delegate callbacks. In that case, you need to
+	//       rewrite the type to have delegate* unmanaged<> *AND* change the callback
+	//       implementation to have [UnmanagedCallersOnly]
+	//
+	//    6. arrays. These are uncommon and fall into two broad categories:
```
Comment (haritha-mohan) [inline tests/cecil-tests/BlittablePInvokes.cs:115]:
```suggestion
	//    5. arrays. These are uncommon and fall into two broad categories:
```

**Review by haritha-mohan — APPROVED** [review]

**Review by mandel-macaque — APPROVED** [review]

**Pull request merged**